integration-tests/src/tests/client/block_corruption.rs

```diff
@@ -99,3 +104,175 @@ fn change_shard_id_to_invalid() {
+///
+/// Returns `Ok(processing_error)` if corrupt block was parsed and its processing failed.
+/// Returns `Ok(reason)` if corrupt block wasn't parsed or had changes that are not breaking by design.
+/// Returns `Err(reason)` if corrupt block was processed or correct block wasn't processed afterwards.
+fn check_process_flipped_block_fails_on_bit(
+    corrupted_bit_idx: usize,
+) -> Result<anyhow::Error, anyhow::Error> {
+    let epoch_length = 5000000;
+    let mut genesis = Genesis::test(vec!["test0".parse().unwrap(), "test1".parse().unwrap()], 1);
+    genesis.config.epoch_length = epoch_length;
+    let mut env = TestEnv::builder(ChainGenesis::new(&genesis))
+        .real_epoch_managers(&genesis.config)
+        .nightshade_runtimes(&genesis)
+        .build();
+
+    let mut last_block = env.clients[0].chain.get_block_by_height(0).unwrap();
+
+    let mid_height = 3;
+    for h in 1..=mid_height {
+        let txs = create_tx_load(h, &last_block);
+        for tx in txs {
+            assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+        }
+
+        let block = env.clients[0].produce_block(h).unwrap().unwrap();
+        env.process_block(0, block.clone(), Provenance::PRODUCED);
+        last_block = block;
+    }
+
+    let block_len = borsh::to_vec(&last_block).unwrap().len();
+    if corrupted_bit_idx >= block_len * 8 {
+        return Ok(anyhow::anyhow!("End"));
+    }
+
+    let h = mid_height + 1;
+
+    let txs = create_tx_load(h, &last_block);
+    for tx in txs {
+        assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+    }
+
+    let correct_block = env.clients[0]
+        .produce_block(h)
+        .unwrap()
+        .with_context(|| format!("Failed to produce block at height {:?}", h))
+        .unwrap();
+
+    let mut block_vec = borsh::to_vec(&correct_block).unwrap();
+
+    // Technically we don't have to do first step, because of Ok("End") check.
+    // But we may change that part in the future for some reason, and I want this part to be reliable.
+
+    // get BIT index modulo BIT length of block
+    let mod_idx = corrupted_bit_idx % (block_vec.len() * 8);
+    // get index of BYTE that we are changing
+    let byte_idx = mod_idx / 8;
+    // get index of BIT in that BYTE that we are flipping
+    let bit_idx = mod_idx % 8;
+    // flip that BIT
+    block_vec[byte_idx] ^= 1 << bit_idx;
+
+    if let Ok(mut corrupt_block) = Block::try_from_slice(block_vec.as_slice()) {
+        corrupt_block.mut_header().get_mut().inner_rest.block_body_hash =
+            corrupt_block.compute_block_body_hash().unwrap();
+        corrupt_block.mut_header().resign(&InMemoryValidatorSigner::from_seed(
+            "test0".parse().unwrap(),
+            KeyType::ED25519,
+            "test0",
+        ));
+
+        if is_breaking_block_change(&correct_block, &corrupt_block) {
```

Comment:
nit: Maybe inverse the condition and return early. This way this method will be less nested and easier to read.